File: Sources/Nice/State/SessionStore.swift
//
//  SessionStore.swift
//  Nice
//
//  Persists the list of live tabs per window to
//  `~/Library/Application Support/Nice/sessions.json` so a relaunch
//  can pick up where the user left off. Claude tabs resume via
//  `claude --resume <uuid>`; terminal-only tabs (including every tab
//  in the pinned Terminals group) restore with a fresh shell.
//
//  Window identity is preserved across quits via
//  `@SceneStorage("windowSessionId")` on `AppShellView`.
//
//  Writes are debounced (500ms) so rapid-fire state mutations during
//  normal use don't thrash the disk, and the encode + atomic write
//  always runs on a private serial `ioQueue` so main is never the
//  writer. `flush()` cancels the pending timer and blocks the caller
//  on a semaphore until the off-main write completes; per-window-close
//  and `willTerminate` both reach `flush()` through `WindowSession.tearDown`,
//  so the last state makes it to disk before the close/exit returns.
//
//  Disk I/O is funneled through `SessionStoreIO` (production:
//  `DiskSessionStoreIO`). Tests inject a recorder to capture writes
//  with queue/thread context — the protocol is the only path from
//  `SessionStore` to the filesystem.
//

import Foundation

struct PersistedPane: Codable, Hashable, Sendable {
    let id: String
    let title: String
    let kind: PaneKind
    /// Last-observed cwd for this pane, captured from OSC 7 in the
    /// injected zsh `chpwd_functions` hook. Optional so v3 session
    /// files (written before per-pane cwd existed) still decode —
    /// restore falls back to `PersistedTab.cwd` when nil.
    let cwd: String?

    init(id: String, title: String, kind: PaneKind, cwd: String? = nil) {
        self.id = id
        self.title = title
        self.kind = kind
        self.cwd = cwd
    }
}

struct PersistedTab: Codable, Hashable, Sendable {
    let id: String
    let title: String
    let cwd: String
    let branch: String?
    /// Non-nil for Claude tabs (used for `claude --resume <uuid>` on
    /// restore). Nil for terminal-only tabs — these come back as a
    /// fresh shell in the persisted cwd.
    let claudeSessionId: String?
    let activePaneId: String?
    let panes: [PersistedPane]
    /// Whether the user renamed this tab via the sidebar inline
    /// editor. Optional so v3 session files (written before this flag
    /// existed) still decode — callers hydrate with `?? false`.
    let titleManuallySet: Bool?
    /// Monotonic per-tab counter for auto-naming new terminal panes
    /// "Terminal N". Optional so older session files still decode —
    /// `WindowSession.addRestoredTabModel` recomputes the counter from
    /// existing pane titles when this is nil.
    let nextTerminalIndex: Int?

    init(
        id: String,
        title: String,
        cwd: String,
        branch: String?,
        claudeSessionId: String?,
        activePaneId: String?,
        panes: [PersistedPane],
        titleManuallySet: Bool? = nil,
        nextTerminalIndex: Int? = nil
    ) {
        self.id = id
        self.title = title
        self.cwd = cwd
        self.branch = branch
        self.claudeSessionId = claudeSessionId
        self.activePaneId = activePaneId
        self.panes = panes
        self.titleManuallySet = titleManuallySet
        self.nextTerminalIndex = nextTerminalIndex
    }
}

/// Sidebar project grouping. Preserves the name/path the user actually
/// saw in the sidebar across relaunches — deriving it fresh from each
/// tab's cwd on restore would split a multi-worktree project like
/// "NICE" into one per worktree dir (no common cwd prefix between
/// worktrees).
struct PersistedProject: Codable, Hashable, Sendable {
    let id: String
    let name: String
    let path: String
    let tabs: [PersistedTab]
}

struct PersistedWindow: Codable, Hashable, Sendable {
    let id: String
    let activeTabId: String?
    let sidebarCollapsed: Bool
    let projects: [PersistedProject]
}

extension PersistedWindow {
    /// Total saved tabs across all projects in this window. Used by
    /// callers that want to know "does this window actually have any
    /// restorable state" without caring which project owns what.
    var totalTabCount: Int {
        projects.reduce(0) { $0 + $1.tabs.count }
    }
}

struct PersistedState: Codable, Hashable, Sendable {
    let version: Int
    let windows: [PersistedWindow]

    /// Bumped to 3 when the pinned Terminals group started persisting
    /// terminal-only tabs (making `PersistedTab.claudeSessionId`
    /// optional and dropping the `mainTerminalCwd` window-level
    /// field). v1/v2 files fail to decode and we start fresh; users
    /// with older sessions.json lose saved tabs on first launch of
    /// this build. Acceptable one-off migration, mirroring the v1→v2
    /// bump.
    static let currentVersion = 3
    static let empty = PersistedState(version: currentVersion, windows: [])
}

/// Surface used by `WindowSession` for persistence — a protocol so
/// unit tests can swap in `FakeSessionStore` to capture upserts /
/// flushes without touching `~/Library/Application Support/Nice/`.
/// `SessionStore.shared` is the only production implementer.
@MainActor
protocol SessionStorePersisting: AnyObject {
    func load() -> PersistedState
    func upsert(window: PersistedWindow)
    func pruneEmptyWindows(keeping: String)
    /// Cancel any pending debounced write and persist the current
    /// state. Conformers must block until the write completes so
    /// callers (`willTerminate`, per-window-close) can rely on the
    /// state being on disk before the call returns.
    func flush()
}

/// Disk side-effects performed by `SessionStore`. Carved out as a
/// protocol so tests can inject a recorder that captures writes
/// (with queue/thread context) and skip touching disk, and so the
/// only path from `SessionStore` to the filesystem is one method
/// call away from the I/O queue. Production uses
/// `DiskSessionStoreIO`.
protocol SessionStoreIO: Sendable {
    /// Encode `state` and write it to `url` atomically. Always
    /// invoked on `SessionStore`'s I/O queue — implementations must
    /// be safe to call from a background thread.
    func write(_ state: PersistedState, to url: URL)
    /// Synchronously read the persisted state at `url`, returning
    /// `.empty` if the file is missing or fails to decode. Called
    /// once at `SessionStore.init`.
    func read(from url: URL) -> PersistedState
}

/// Production `SessionStoreIO`: JSON-encode → atomic write to disk
/// with `try?`. A failed atomic write leaves the prior file intact,
/// which is the recovery contract pinned by
/// `SessionStoreWriteFailureTests`.
struct DiskSessionStoreIO: SessionStoreIO {
    func write(_ state: PersistedState, to url: URL) {
        let encoder = JSONEncoder()
        // `.withoutEscapingSlashes` keeps path values readable
        // (`/Users/...` instead of `\/Users\/...`). Both forms
        // decode to the same string, so this is purely cosmetic.
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        guard let data = try? encoder.encode(state) else { return }
        try? data.write(to: url, options: .atomic)
    }

    func read(from url: URL) -> PersistedState {
        guard let data = try? Data(contentsOf: url) else { return .empty }
        guard let decoded = try? JSONDecoder().decode(PersistedState.self, from: data) else {
            return .empty
        }
        return decoded
    }
}

@MainActor
final class SessionStore: SessionStorePersisting {
    /// Label of the serial queue that owns every `SessionStore` write.
    /// Exposed (internal) so tests can assert "this write ran on
    /// `ioQueue`, not main, not some other queue" via the dispatch
    /// queue label C API.
    static let ioQueueLabel = "dev.nickanderssohn.nice.sessionstore.io"

    static let shared = SessionStore()

    private let fileURL: URL
    private var pendingWork: DispatchWorkItem?
    /// Debounce window for `upsert`. Short enough that a quick ⌘W +
    /// ⌘Q still catches the final state (the subsequent `flush` in
    /// `tearDown` runs synchronously and cancels the pending work).
    private let debounceInterval: TimeInterval = 0.5

    /// Serial background queue for the encode + atomic file write.
    /// Keeps disk I/O off the main thread; serial so a flush write
    /// always runs after any earlier debounced write that already
    /// left the main-thread timer. Note: `qos: .utility` plus a
    /// main-thread `sem.wait()` in `flush()` is technically a
    /// priority inversion, but macOS auto-promotes the queue's QoS
    /// for the duration of the wait, so the wait isn't starved.
    private let ioQueue = DispatchQueue(
        label: SessionStore.ioQueueLabel,
        qos: .utility
    )

    /// Disk-side effects. Defaults to `DiskSessionStoreIO` in
    /// production; tests inject a recorder so they can assert on
    /// writes (count, ordering, queue) without touching disk.
    private let io: SessionStoreIO

    /// In-memory cache of the last-read / last-written state. Keeps
    /// `upsert` cheap: it mutates the cached struct and schedules a
    /// single write instead of re-reading the file on every change.
    private var cached: PersistedState

    init(io: SessionStoreIO = DiskSessionStoreIO()) {
        self.io = io
        let fm = FileManager.default
        // Tests set `NICE_APPLICATION_SUPPORT_ROOT` to redirect session
        // state into a sandbox directory. `FileManager.url(for:
        // .applicationSupportDirectory, ...)` resolves the user's home
        // via `getpwuid(getuid())` and ignores `$HOME`, so overriding
        // HOME alone isn't enough to keep UITests from reading and
        // mutating the user's real `~/Library/Application Support/Nice/
        // sessions.json`. Production leaves this unset.
        let root: URL
        if let override = ProcessInfo.processInfo.environment["NICE_APPLICATION_SUPPORT_ROOT"],
           !override.isEmpty {
            root = URL(fileURLWithPath: override, isDirectory: true)
        } else {
            // `create: true` already creates the directory; we still
            // append the folder name ourselves and create that subdir
            // below.
            root = (try? fm.url(
                for: .applicationSupportDirectory,
                in: .userDomainMask,
                appropriateFor: nil,
                create: true
            )) ?? URL(fileURLWithPath: NSHomeDirectory())
                .appendingPathComponent("Library/Application Support", isDirectory: true)
        }
        // Folder name tracks CFBundleName so the `Nice Dev` variant
        // lands at `~/Library/Application Support/Nice Dev/` and can't
        // clobber the user's real sessions in `…/Nice/`.
        let folder = (Bundle.main.object(forInfoDictionaryKey: "CFBundleName") as? String) ?? "Nice"
        let supportDir = root.appendingPathComponent(folder, isDirectory: true)
        try? fm.createDirectory(at: supportDir, withIntermediateDirectories: true)
        self.fileURL = supportDir.appendingPathComponent("sessions.json")
        self.cached = io.read(from: fileURL)
    }

    /// Return the current persisted state. Safe to call at any time —
    /// returns the cached value without hitting disk after the first
    /// read.
    func load() -> PersistedState {
        return cached
    }

    /// Merge `window` into the persisted state, replacing any existing
    /// entry with the same id. Schedules a debounced write; callers
    /// don't need to block on I/O.
    func upsert(window: PersistedWindow) {
        var windows = cached.windows
        if let idx = windows.firstIndex(where: { $0.id == window.id }) {
            windows[idx] = window
        } else {
            windows.append(window)
        }
        cached = PersistedState(version: PersistedState.currentVersion, windows: windows)
        scheduleWrite()
    }

    /// Drop every window with zero tabs except `keep` (usually the
    /// window id of the caller so they can still save into their
    /// slot). Called from `AppState.restoreSavedWindow` to garbage-
    /// collect "ghost" entries accumulated by prior launches whose
    /// restore attempts all failed.
    func pruneEmptyWindows(keeping keep: String) {
        let filtered = cached.windows.filter { $0.id == keep || $0.totalTabCount > 0 }
        guard filtered.count != cached.windows.count else { return }
        cached = PersistedState(version: PersistedState.currentVersion, windows: filtered)
        scheduleWrite()
    }

    /// Cancel any pending debounced write and flush the cache to disk.
    /// The write itself runs on `ioQueue` (so main is never the
    /// writer); the semaphore makes this call block until the write
    /// completes, preserving the "flushed before terminate returns"
    /// guarantee. Called from `WindowSession.tearDown` on per-window
    /// close *and* on app terminate.
    func flush() {
        dispatchPrecondition(condition: .onQueue(.main))
        pendingWork?.cancel()
        pendingWork = nil
        enqueueWriteAndWait(snapshot: cached)
    }

    // MARK: - Disk I/O

    private func scheduleWrite() {
        dispatchPrecondition(condition: .onQueue(.main))
        pendingWork?.cancel()
        let snapshot = cached
        // Timer still lives on main so cancel-on-rescheduling stays
        // race-free with `pendingWork` (a `@MainActor`-isolated ivar).
        // When it fires, it hands the snapshot to `enqueueWrite`; the
        // actual encode + atomic write runs on the background worker.
        let work = DispatchWorkItem { [weak self] in
            MainActor.assumeIsolated {
                self?.enqueueWrite(snapshot: snapshot)
            }
        }
        pendingWork = work
        DispatchQueue.main.asyncAfter(
            deadline: .now() + debounceInterval, execute: work
        )
    }

    /// Single funnel for non-blocking writes. Captures `io` and `url`
    /// as values, then dispatches to `ioQueue`. The
    /// `dispatchPrecondition` inside the block is the structural guard
    /// that no future caller bypasses the I/O queue: every write goes
    /// through here.
    private func enqueueWrite(snapshot: PersistedState) {
        let io = self.io
        let url = self.fileURL
        ioQueue.async { [ioQueue] in
            dispatchPrecondition(condition: .onQueue(ioQueue))
            io.write(snapshot, to: url)
        }
    }

    /// Blocking variant for `flush`. Same enqueue path, plus a
    /// semaphore so the caller can rely on the write being on disk
    /// before the call returns.
    private func enqueueWriteAndWait(snapshot: PersistedState) {
        let io = self.io
        let url = self.fileURL
        let sem = DispatchSemaphore(value: 0)
        ioQueue.async { [ioQueue] in
            dispatchPrecondition(condition: .onQueue(ioQueue))
            io.write(snapshot, to: url)
            sem.signal()
        }
        sem.wait()
    }
}

File: crates/nice/src/session_store.rs
//! Session store (R18) — ports Swift `SessionStore` (`SessionStore.swift`).
//!
//! Persists the per-window tab tree to
//! `<app-support>/Nice RS Dev/sessions.json` so a relaunch restores
//! windows/tabs/panes. Claude tabs resume via `claude --resume <uuid>`;
//! terminal-only tabs restore with a fresh shell in their saved cwd.
//!
//! ## Schema
//!
//! The window envelope (`PersistedState`/`PersistedWindow`/`PersistedFrame`)
//! lives here; the model-shaped leaves (`PersistedPane`/`PersistedTab`/
//! `PersistedProject`) live gpui-free in `nice-model`. The schema is Swift's v3
//! **minus `branch`** (M5), tolerant by SHAPE: no version gate on read, unknown
//! fields ignored (NO `deny_unknown_fields`), nil-omitted optionals. A
//! missing/corrupt/shape-mismatched file decodes to `{version:3, windows:[]}` —
//! never an error, so the app always launches. Writes are `version: 3`.
//!
//! ## Store machinery (the observable contract — dossier §3.2)
//!
//! * (a) mutations never block — `upsert`/`remove`/`prune_empty_windows` update
//!   the in-memory cache and schedule a debounced write;
//! * (b) 500 ms debounce coalescing (each mutation pushes the deadline out);
//! * (c) `flush()` is synchronous — cancels the pending timer, forces a write,
//!   and blocks until it lands;
//! * (d) the writer never runs concurrently with itself — ONE dedicated OS
//!   thread owns every write (the `control_socket` thread precedent; gpui
//!   timers are App-Nap-deferred, so a scheduler-level thread event, not a
//!   parked timer, drives the cadence);
//! * (e) atomic write (temp + rename via the shared [`crate::atomic_file`]
//!   helper) — a failed write leaves the prior file intact;
//! * (f) `remove(id)` / `prune_empty_windows(keeping)` filter the cache.
//!
//! ## Global + absent ⇒ no-op
//!
//! The store lives in a process [`GLOBAL`] installed by `app::run` only. Every
//! persistence hook goes through the [`upsert`]/[`remove`]/[`prune`]/[`flush`]/
//! [`load`] free functions, which are a **no-op when the Global is absent** —
//! scenarios/tests opt in via an injected temp store path
//! ([`install_global`]). Nothing here reads or writes the developer's real
//! `~/Library/Application Support`; the base dir is injectable
//! (`NICE_APPLICATION_SUPPORT_ROOT`, resolved only in `app::run`).
//!
//! ## Migration
//!
//! One-time: iff the OWN file is ABSENT, [`open`] reads the Swift app's
//! `…/Nice/sessions.json` (source path injectable) with the same tolerance,
//! drops `branch` (the leaf structs have no such field), and writes to the OWN
//! store only — writing the Swift path would fight a still-installed Swift
//! Nice's flushes.
//!
//! The launch-bootstrap + restore-fan-out consumers land with a later slice,
//! hence the module-wide `dead_code` allow (the established
//! later-slice-consumer pattern).

#![allow(dead_code)]

use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Condvar, Mutex, OnceLock};
use std::thread::JoinHandle;
use std::time::{Duration, Instant};

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use nice_model::PersistedProject;

/// The schema version R18 writes. v1/v2 files fail to decode and start fresh
/// (the same one-off migration Swift accepted at the v2→v3 bump).
pub const CURRENT_VERSION: i64 = 3;

/// Debounce window for `upsert`/`remove`/`prune`. Short enough that a quick ⌘W
/// + ⌘Q still catches the final state (the subsequent synchronous `flush`
/// cancels the pending timer).
pub const DEBOUNCE: Duration = Duration::from_millis(500);

/// Application-support subfolder for the Rust build — deliberately distinct
/// from the Swift `Nice` / `Nice Dev` folders so it can't clobber the user's
/// real sessions.
pub const STORE_FOLDER: &str = "Nice RS Dev";

/// The Swift app's application-support subfolder — the one-time migration
/// SOURCE. Read only when the own store is absent; NEVER written.
pub const SWIFT_STORE_FOLDER: &str = "Nice";

/// On-screen frame captured at last save (Cocoa window coordinates, origin at
/// the bottom-left of the primary screen — matches Swift `PersistedFrame`, so
/// migration needs no value conversion). Raw `f64`s keep the JSON shape stable
/// and human-readable.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistedFrame {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

/// One persisted window. `sidebar_collapsed` is REQUIRED (Swift's decode
/// requires it); `frame` is optional so pre-frame-persistence files decode and
/// fall back to default placement.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistedWindow {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_tab_id: Option<String>,
    pub sidebar_collapsed: bool,
    pub projects: Vec<PersistedProject>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frame: Option<PersistedFrame>,
}

impl PersistedWindow {
    /// Total saved tabs across all projects — "does this window have restorable
    /// state" without caring which project owns what.
    pub fn total_tab_count(&self) -> usize {
        self.projects.iter().map(|p| p.tabs.len()).sum()
    }
}

/// The whole persisted document. `version` is written `3`; unknown future
/// versions still decode (forward-compat) since there is no version gate.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PersistedState {
    pub version: i64,
    pub windows: Vec<PersistedWindow>,
}

impl PersistedState {
    /// The empty state a missing/corrupt/shape-mismatched file decodes to.
    pub fn empty() -> Self {
        PersistedState {
            version: CURRENT_VERSION,
            windows: Vec::new(),
        }
    }
}

// MARK: - Disk I/O seam

/// Disk side-effects performed by [`SessionStore`]. A trait so tests inject a
/// recorder that captures writes (with thread context) and skips touching disk.
/// Production uses [`DiskIo`]. Always invoked on the store's dedicated writer
/// thread — implementations must be safe off the calling thread.
pub trait SessionStoreIo: Send + Sync + 'static {
    fn write(&self, state: &PersistedState, path: &Path);
}

/// Production `SessionStoreIo`: serialize pretty + sorted-keys → atomic write.
/// A failed atomic write leaves the prior file intact (the write-failure
/// recovery contract).
pub struct DiskIo;

impl SessionStoreIo for DiskIo {
    fn write(&self, state: &PersistedState, path: &Path) {
        if let Ok(bytes) = serialize_state(state) {
            // `try?` semantics: a failed write is swallowed; the atomic rename
            // never ran, so the prior file is untouched.
            let _ = crate::atomic_file::write_atomic(path, &bytes, None);
        }
    }
}

/// Serialize `state` pretty-printed with recursively sorted keys (Swift's
/// `.sortedKeys`), so the output is byte-stable regardless of serde_json's
/// `preserve_order` feature — the property the write-if-changed compare needs.
fn serialize_state(state: &PersistedState) -> serde_json::Result<Vec<u8>> {
    let value = serde_json::to_value(state)?;
    serde_json::to_vec_pretty(&sort_value(value))
}

/// Recursively rebuild `v` with every object's keys in sorted order (arrays
/// keep their element order).
fn sort_value(v: Value) -> Value {
    match v {
        Value::Object(map) => {
            let mut keys: Vec<String> = map.keys().cloned().collect();
            keys.sort();
            let mut sorted = Map::new();
            for k in keys {
                let child = map.get(&k).cloned().unwrap_or(Value::Null);
                sorted.insert(k, sort_value(child));
            }
            Value::Object(sorted)
        }
        Value::Array(items) => Value::Array(items.into_iter().map(sort_value).collect()),
        other => other,
    }
}

/// Read + tolerantly decode the state at `path`. Missing / empty / corrupt /
/// shape-mismatched ⇒ [`PersistedState::empty`], never an error (Swift's
/// `read(from:)` contract).
pub fn read_state(path: &Path) -> PersistedState {
    let Ok(bytes) = fs::read(path) else {
        return PersistedState::empty();
    };
    if bytes.is_empty() {
        return PersistedState::empty();
    }
    serde_json::from_slice(&bytes).unwrap_or_else(|_| PersistedState::empty())
}

// MARK: - The store

/// Writer-thread coordination state, behind one mutex + condvar.
struct Inner {
    /// Last-read / last-written state — the in-memory cache.
    cached: PersistedState,
    /// Bumped on every cache mutation. `flush` blocks until the writer reports
    /// having written this revision.
    revision: u64,
    /// The last revision the writer has written (attempted). `== revision`
    /// means the cache is on disk.
    written_revision: u64,
    /// When the pending debounced write should fire, if one is scheduled.
    deadline: Option<Instant>,
    /// Force an immediate write of the current revision (set by `flush`).
    flush_now: bool,
    /// Writer-thread exit signal.
    shutdown: bool,
}

struct Shared {
    path: PathBuf,
    io: Box<dyn SessionStoreIo>,
    debounce: Duration,
    inner: Mutex<Inner>,
    cv: Condvar,
}

/// A live session store: an in-memory cache plus one dedicated writer thread.
/// Dropping the store flushes anything pending and joins the writer, so no
/// thread leaks.
pub struct SessionStore {
    shared: Arc<Shared>,
    handle: Option<JoinHandle<()>>,
}

impl SessionStore {
    /// Open the store at `own_path`, running the one-time Swift migration when
    /// the own file is absent and `swift_source` is given. Production
    /// (`DiskIo`, 500 ms debounce).
    pub fn open(own_path: PathBuf, swift_source: Option<PathBuf>) -> Self {
        Self::open_with(own_path, swift_source, Box::new(DiskIo), DEBOUNCE)
    }

    /// [`SessionStore::open`] with an injected I/O seam + debounce (tests).
    pub fn open_with(
        own_path: PathBuf,
        swift_source: Option<PathBuf>,
        io: Box<dyn SessionStoreIo>,
        debounce: Duration,
    ) -> Self {
        // Own file present ⇒ read it. Absent ⇒ migrate from the Swift source
        // (branch auto-dropped: our leaf structs have no `branch` field). The
        // Swift file is only ever READ.
        let (cached, migrated) = if own_path.exists() {
            (read_state(&own_path), false)
        } else if let Some(src) = &swift_source {
            (read_state(src), true)
        } else {
            (PersistedState::empty(), false)
        };

        let store = Self::build(own_path, cached, io, debounce);

        // One-time migration write: persist the adopted state to the OWN store
        // so a subsequent launch reads its own file (and never re-reads a
        // since-changed Swift file). Never touches `swift_source`.
        if migrated && !store.shared.inner.lock().unwrap().cached.windows.is_empty() {
            {
                let mut inner = store.shared.inner.lock().unwrap();
                inner.revision += 1;
            }
            store.flush();
        }

        store
    }

    /// Construct the store + spawn its writer thread. Creates the parent dir of
    /// `path` best-effort so the first write lands.
    fn build(
        path: PathBuf,
        cached: PersistedState,
        io: Box<dyn SessionStoreIo>,
        debounce: Duration,
    ) -> Self {
        if let Some(parent) = path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let shared = Arc::new(Shared {
            path,
            io,
            debounce,
            inner: Mutex::new(Inner {
                cached,
                revision: 0,
                written_revision: 0,
                deadline: None,
                flush_now: false,
                shutdown: false,
            }),
            cv: Condvar::new(),
        });
        let handle = {
            let shared = Arc::clone(&shared);
            std::thread::Builder::new()
                .name("nice-session-store-writer".into())
                .spawn(move || writer_loop(shared))
                .expect("spawn session-store writer thread")
        };
        SessionStore {
            shared,
            handle: Some(handle),
        }
    }

    /// Return the current persisted state (the cache; no disk hit).
    pub fn load(&self) -> PersistedState {
        self.shared.inner.lock().unwrap().cached.clone()
    }

    /// Merge `window` into the cache, replacing any entry with the same id, and
    /// schedule a debounced write. Never blocks.
    pub fn upsert(&self, window: PersistedWindow) {
        let mut inner = self.shared.inner.lock().unwrap();
        let mut windows = std::mem::take(&mut inner.cached.windows);
        match windows.iter().position(|w| w.id == window.id) {
            Some(idx) => windows[idx] = window,
            None => windows.push(window),
        }
        inner.cached = PersistedState {
            version: CURRENT_VERSION,
            windows,
        };
        self.mark_dirty_and_schedule(&mut inner);
    }

    /// Drop the entry whose id matches. No-op (no write) if the id isn't
    /// present — a quit right after a `remove` must not resurrect the slot from
    /// a stale debounce, and a no-op must not schedule spurious I/O.
    pub fn remove(&self, window_id: &str) {
        let mut inner = self.shared.inner.lock().unwrap();
        let before = inner.cached.windows.len();
        let windows: Vec<PersistedWindow> = std::mem::take(&mut inner.cached.windows)
            .into_iter()
            .filter(|w| w.id != window_id)
            .collect();
        if windows.len() == before {
            inner.cached.windows = windows;
            return;
        }
        inner.cached = PersistedState {
            version: CURRENT_VERSION,
            windows,
        };
        self.mark_dirty_and_schedule(&mut inner);
    }

    /// Drop every window with zero tabs except `keeping` (the caller's own slot
    /// so it can still save into it). No-op (no write) when nothing changes.
    pub fn prune_empty_windows(&self, keeping: &str) {
        self.prune_empty_windows_keeping(&[keeping.to_string()]);
    }

    /// [`prune_empty_windows`](Self::prune_empty_windows) keeping a SET of ids —
    /// the R18 restore fan-out's post-restore GC (Swift `pruneEmptyWindows(keeping:)`
    /// run after adoption), keeping every just-restored window id so a legitimately
    /// empty Terminals-only restored window survives the prune. No-op (no write)
    /// when nothing changes.
    pub fn prune_empty_windows_keeping(&self, keeping: &[String]) {
        let mut inner = self.shared.inner.lock().unwrap();
        let before = inner.cached.windows.len();
        let windows: Vec<PersistedWindow> = std::mem::take(&mut inner.cached.windows)
            .into_iter()
            .filter(|w| keeping.iter().any(|k| k == &w.id) || w.total_tab_count() > 0)
            .collect();
        if windows.len() == before {
            inner.cached.windows = windows;
            return;
        }
        inner.cached = PersistedState {
            version: CURRENT_VERSION,
            windows,
        };
        self.mark_dirty_and_schedule(&mut inner);
    }

    fn mark_dirty_and_schedule(&self, inner: &mut Inner) {
        inner.revision += 1;
        inner.deadline = Some(Instant::now() + self.shared.debounce);
        self.shared.cv.notify_all();
    }

    /// Cancel any pending debounced write and flush the cache to disk
    /// synchronously — blocks until the write completes, preserving the
    /// "flushed before terminate returns" guarantee. Writes only when the cache
    /// is ahead of disk (nothing to flush ⇒ returns immediately).
    pub fn flush(&self) {
        let mut inner = self.shared.inner.lock().unwrap();
        let target = inner.revision;
        if inner.written_revision >= target {
            return;
        }
        inner.flush_now = true;
        inner.deadline = None; // flush cancels/coalesces the pending debounce
        self.shared.cv.notify_all();
        while inner.written_revision < target {
            inner = self.shared.cv.wait(inner).unwrap();
        }
    }
}

impl Drop for SessionStore {
    fn drop(&mut self) {
        // Persist anything pending, then stop + join the writer (no thread
        // leak, no lost write).
        self.flush();
        {
            let mut inner = self.shared.inner.lock().unwrap();
            inner.shutdown = true;
        }
        self.shared.cv.notify_all();
        if let Some(handle) = self.handle.take() {
            let _ = handle.join();
        }
    }
}

/// The dedicated writer thread: sleeps until a debounced deadline elapses, a
/// flush forces a write, or shutdown. Serial by construction — the ONLY writer.
fn writer_loop(shared: Arc<Shared>) {
    loop {
        let mut inner = shared.inner.lock().unwrap();
        // Wait for work.
        loop {
            if inner.shutdown && inner.deadline.is_none() && !inner.flush_now {
                return;
            }
            if inner.flush_now {
                break;
            }
            match inner.deadline {
                Some(deadline) => {
                    let now = Instant::now();
                    if now >= deadline {
                        break;
                    }
                    let (guard, _) = shared.cv.wait_timeout(inner, deadline - now).unwrap();
                    inner = guard;
                }
                None => {
                    if inner.shutdown {
                        return;
                    }
                    inner = shared.cv.wait(inner).unwrap();
                }
            }
        }

        let target = inner.revision;
        let force = inner.flush_now;
        inner.flush_now = false;
        inner.deadline = None;
        if !force && inner.written_revision >= target {
            // Debounced timer fired with nothing new to write.
            continue;
        }
        let snapshot = inner.cached.clone();
        drop(inner);

        shared.io.write(&snapshot, &shared.path);

        let mut inner = shared.inner.lock().unwrap();
        inner.written_revision = target;
        shared.cv.notify_all();
    }
}

// MARK: - Path resolution (called from app::run ONLY)

/// Application-support root: `NICE_APPLICATION_SUPPORT_ROOT` when set (tests /
/// scenarios redirect state into a sandbox), else `~/Library/Application
/// Support`. Resolved only in `app::run` (the `shell_inject.rs` convention) —
/// never in a test or `run_selftest`.
fn support_root() -> PathBuf {
    match env::var("NICE_APPLICATION_SUPPORT_ROOT") {
        Ok(v) if !v.is_empty() => PathBuf::from(v),
        _ => {
            let home = env::var("HOME").unwrap_or_else(|_| "/".to_string());
            PathBuf::from(home).join("Library/Application Support")
        }
    }
}

/// The OWN store path: `<support-root>/Nice RS Dev/sessions.json`.
pub fn default_store_path() -> PathBuf {
    support_root().join(STORE_FOLDER).join("sessions.json")
}

/// The Swift migration SOURCE path: `<support-root>/Nice/sessions.json`. Read
/// only when the own store is absent; never written.
pub fn swift_migration_source() -> PathBuf {
    support_root().join(SWIFT_STORE_FOLDER).join("sessions.json")
}

// MARK: - Process Global (absent ⇒ every persistence hook is a no-op)

fn global_cell() -> &'static Mutex<Option<Arc<SessionStore>>> {
    static GLOBAL: OnceLock<Mutex<Option<Arc<SessionStore>>>> = OnceLock::new();
    GLOBAL.get_or_init(|| Mutex::new(None))
}

/// Install `store` as the process store, returning the shared handle. Called by
/// `app::run` only.
pub fn install_global(store: SessionStore) -> Arc<SessionStore> {
    let arc = Arc::new(store);
    *global_cell().lock().unwrap() = Some(Arc::clone(&arc));
    arc
}

/// The installed store, if any.
pub fn global() -> Option<Arc<SessionStore>> {
    global_cell().lock().unwrap().clone()
}

/// Uninstall the process store (test teardown / shutdown). Dropping the last
/// `Arc` flushes + joins the writer.
pub fn clear_global() {
    *global_cell().lock().unwrap() = None;
}

/// Persistence hooks — each a NO-OP when no store is installed.
pub fn upsert(window: PersistedWindow) {
    if let Some(store) = global() {
        store.upsert(window);
    }
}

pub fn remove(window_id: &str) {
    if let Some(store) = global() {
        store.remove(window_id);
    }
}

pub fn prune_empty_windows(keeping: &str) {
    if let Some(store) = global() {
        store.prune_empty_windows(keeping);
    }
}

pub fn prune_empty_windows_keeping(keeping: &[String]) {
    if let Some(store) = global() {
        store.prune_empty_windows_keeping(keeping);
    }
}

pub fn flush() {
    if let Some(store) = global() {
        store.flush();
    }
}

/// The current persisted state — empty when no store is installed.
pub fn load() -> PersistedState {
    global()
        .map(|store| store.load())
        .unwrap_or_else(PersistedState::empty)
}

#[cfg(test)]
mod tests;
